### PR TITLE
wta(tui): queue prompts while busy, Esc to dequeue LIFO

### PR DESCRIPTION
**Pull request by yeelam-gordon**
## Summary

Adds a per-tab pending-prompt queue to the WTA TUI. Today, pressing Enter while a turn is in flight surfaces a "Agent is busy on this tab — wait for the current prompt to finish." system message and **discards the user's typing**. With this change:

- **Enter while busy** → push the prompt onto the tab's `pending_prompts` queue and clear the input box.
- **FIFO drain** → when the turn returns to an accepting state (`accepts_new_prompt() && recommendations().is_none() && !loading_session`) the front of the queue is dispatched automatically.
- **Esc behavior** with an empty input box, in priority order:
  1. Dismiss any visible overlay (help / notifications / recommendation card / suggested-pane indicator).
  2. **Pop the back of `pending_prompts` (LIFO undo)** — if the queue is non-empty, Esc rescinds the most-recently queued prompt. The in-flight head (whether autofix or a user turn) keeps running. Repeated Esc presses peel queued prompts off the back one by one.
  3. **Cancel the in-flight head** like `/stop` — only fires once the queue is empty.
  4. Fallback → clear input draft (existing behavior).
- **Visible indicator** → a 1-row `Queued (N): preview…` rendered directly above the input box.
- **Bounded** at `PENDING_PROMPT_QUEUE_CAP = 20`. Over-cap keeps the user's input intact and shows a transient hint instead of silently dropping it.
- **Queue cleared on reset** — `/clear`, `/new`, `/restart`, tab reset and session-load all funnel through `clear_chat_history()`, which now drops `pending_prompts` so stale prompts can't fire into a fresh session.

## Design notes

- **Why queue-pop beats cancel-head:** in the autofix-with-queue scenario (autofix in flight, user typed a follow-up question), users want a single Esc to rescind their typed question — not tear down the autofix the system started. Queue items represent work the *user* added; the head may have been added by autofix. Peeling from the back first matches the user's mental model of "undo my most recent action".
- **`cancel_in_flight_turn()` helper** shared between Ctrl+C and the new Esc branch. Falls back to `DEFAULT_TAB_ID` when the focused tab has no ACP session id yet — same convention the Enter dispatch path uses, so the local state machine still resets correctly even before `SessionAttached`.
- **Drain hook lives at the end of every event-handling tick** (both `ui_rx` and `event_rx` paths) rather than wired into each individual `TurnState` transition. Error paths (`TabError`, `AgentError`), stale-autofix discards, and the autofix `Ignore` decision are all covered uniformly without scattering callsites.
- **Per-tick at most one dispatch per tab** — `turn_submit_prompt` flips state to `Submitted`, so a second pass would no-op. Per-tab FIFO order is deterministic; cross-tab order is sorted alphabetically for log reproducibility.
- **Card-visible blocks drain** — when a recommendation card is staged (`Surfaced { end_pending: false }`) `accepts_new_prompt()` returns true, but a queued prompt going through `turn_submit_prompt` would clear `selected_recommendation` and wipe the card before the user can Run / Insert / dismiss.
- **Tab-scoped dispatch** — walks `tab_sessions`, resolves session-id per tab. Background tabs without an ACP `session_id` defer drain (avoids the `tab_for_session(DEFAULT_TAB_ID)` → focused-tab misroute).
- **Cached collapsed preview** — `QueuedPrompt::new()` precomputes the whitespace-collapsed text once at enqueue so the per-frame UI render is allocation-free.

## Tests

`cargo test --target x86_64-pc-windows-msvc --manifest-path tools/wta/Cargo.toml` — **303 tests pass** (271 baseline + 18 new queue/cancel/truncation + 14 inherited from #42's autofix-per-tab refactor after merge).

Highlights:
| Test | Covers |
|---|---|
| `enter_while_busy_queues_prompt_without_user_bubble` | Enter mid-flight enqueues; no premature User bubble |
| `drain_dispatches_queued_prompts_fifo_when_turn_completes` | FIFO drain after `turn_cancel` |
| `drain_uses_loop_tab_id_for_pane_context_not_focused_tab` | Background-tab drain attributes to right tab |
| `drain_holds_queue_when_tab_has_no_session_id` | No `DEFAULT_TAB_ID` misroute for background tabs |
| `esc_pops_queue_before_cancelling_in_flight` | Pins the queue-pop-priority contract |
| `esc_cancels_autofix_only_when_queue_is_empty` | Reported autofix-with-queue scenario |
| `esc_cancels_in_flight_turn_when_input_empty` | Empty-queue + in-flight Esc behaves like `/stop` |
| `esc_pops_back_of_queue_when_idle_and_input_empty` | Idle + queue → LIFO pop |
| `esc_does_not_pop_queue_when_input_is_not_empty` | Draft preserved; queue + in-flight preserved |
| `drain_skips_when_recommendation_card_visible` | Card blocks auto-drain |
| `queue_full_keeps_input_and_emits_hint` | Cap enforced; input preserved on overflow |
| `clear_chat_history_drops_pending_prompts` | Reset wipes queue |
| `drain_no_ops_when_disconnected` | Disconnected state defers drain |
| `queued_prompt_preview_truncates_and_collapses_whitespace` | Preview sanitized |
| `queued_prompt_preview_zero_budget_is_empty` | `preview(0)` honors char-budget contract |
| `truncate_respects_width_with_combining_marks` | Width-correct truncation with zero-width chars |
| `truncate_wide_char_with_narrow_budget_emits_ellipsis` | CJK + max_cells=1 → bare `…` |
| `truncate_wide_char_with_two_cell_budget_emits_ellipsis_only` | Width contract under tight budget |

## Files

| File | Change |
|---|---|
| `tools/wta/src/app.rs` | `QueuedPrompt` type with cached `collapsed`, `pending_prompts` field, Enter enqueue, Esc cascade (queue-pop → cancel-in-flight → clear-input), `cancel_in_flight_turn()` helper shared with Ctrl+C, `drain_pending_prompts()`, drain hooks in `run()`, 13 new unit tests |
| `tools/wta/src/ui/queued_hint.rs` | **new** — renders `Queued (N): preview…` row with width-correct truncation (handles CJK, combining marks, narrow panes) |
| `tools/wta/src/ui/layout.rs` | reserve + render queue row above input; cursor positioning replicates render's `chat_estimate.min(chat_max)` clamp + welcome-hint reservation |
| `tools/wta/src/ui/mod.rs` | export `queued_hint` |
| `tools/wta/locales/*.yml` | `input.queue.indicator/removed/full` keys added to **all 89 locales** |

## Out of scope

- **RTL (Arabic/Hebrew) bidirectional reordering** of chat/input rows — separate, larger investigation. (The new queue strings ARE translated for RTL locales; only chat-pane bidi reordering is deferred.)

### CODE REVIEW

File: tools/wta/src/app.rs [hunk 1]
```diff
@@ -5005,6 +5087,87 @@ impl App {
...
+                .tab_sessions
+                .get_mut(&tab_id)
+                .expect("tab present")
+                .pending_prompts
+                .pop_front()
+                .expect("non-empty checked");
+
+            let pane_context = PaneContext {
+                pane_id: self.pane_id.clone(),
+                tab_id: self.tab_id.clone(),
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Good catch — fixed in bc35e4480. drain_pending_prompts now uses the loop's 	ab_id for PaneContext.tab_id (and added the test `drain_uses_loop_tab_id_for_pane_context_not_focused_tab` to pin the contract).

File: tools/wta/src/ui/queued_hint.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        "input.queue.indicator",
+        count = count,
+        preview = preview
+    )
+    .into_owned();
+    // Truncate again at the line level just in case the localized template
+    // expands beyond the available width (e.g. RTL or longer translations).
+    let truncated = truncate_to_width(&text, area.width as usize);
+    let line = Line::from(Span::styled(
+        format!("  {}", truncated),
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Fixed in bc35e4480. Truncation now subtracts `HORIZONTAL_PADDING` from `area.width` before passing to `truncate_to_width` so the prepended `\\\` two-space padding doesn't push localized text off the right edge.

File: tools/wta/src/app.rs [hunk 3]
```diff
@@ -7503,4 +7666,216 @@
...
+        assert_eq!(app.current_tab().pending_prompts[0].text, "q2");
+        assert_eq!(app.current_tab().pending_prompts[1].text, "q3");
+    }
+
+    #[test]
+    fn esc_pops_back_of_queue_when_input_empty() {
+        let mut app = connected_app_with_text("first");
+        press_enter(&mut app);
+        for q in ["a", "b", "c"] {
+            type_text(&mut app, q);
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs:8491]: Fixed in 2e0375747 — renamed to `preempt` in the test assertion.

File: tools/wta/src/app.rs [hunk 4]
```diff
@@ -5005,6 +5087,93 @@ impl App {
...
 
 impl App {
+    /// Walk every tab and dispatch the front of its `pending_prompts` queue
+    /// if the turn state is ready for a new prompt. Idempotent and cheap when
+    /// queues are empty — safe to call after every event-handling tick.
+    ///
+    /// **Per-tick policy:** at most one queued prompt per tab per call.
+    /// `turn_submit_prompt` flips the state to `Submitted`, so a second pass
+    /// would no-op anyway. This keeps ordering deterministic and avoids any
+    /// risk of starving the agent with a synchronous flood.
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Good point. Fixed in 2e0375747: `drain_pending_prompts` now sorts `tab_ids` before iterating, and the doc comment clarifies that **per-tab** FIFO is deterministic while **cross-tab** order is sorted-alphabetical for log reproducibility.

File: tools/wta/src/ui/layout.rs [hunk 5]
```diff
@@ -259,9 +265,10 @@ pub fn input_cursor_position(app: &App, area: Rect) -> Option<Position> {
             Constraint::Min(0),
             hint_height,
             rec_hint_height,
+            queue_hint_height,
             Constraint::Length(input_height),
         ])
         .split(main_area);
 
-    input::cursor_position(app, chunks[6])
+    input::cursor_position(app, chunks[7])
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/layout.rs:287]: Fixed in 2e0375747 — stale comment updated from `index 6` to `index 7` and the constraint sequence now lists `queue_hint` explicitly.

File: tools/wta/src/ui/queued_hint.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    let mut used = 0usize;
+    for ch in text.chars() {
+        let w = UnicodeWidthChar::width(ch).unwrap_or(0);
+        if used + w > max_cells {
+            // Replace the last char with an ellipsis if there's room.
+            if max_cells >= 1 && !out.is_empty() {
+                out.pop();
+                out.push('…');
+            }
+            break;
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs:109]: Good edge-case catch. Fixed in 2e0375747: when `out` is still empty at overflow time (e.g. wide-char first glyph with `max_cells=1`), we now emit a lone `...` so the row is never blank. Regression test `truncate_wide_char_with_narrow_budget_emits_ellipsis` added.

File: dep/vcpkg-overlay-triplets/x64-windows-static.cmake [hunk 7]
```diff
@@ -5,4 +5,4 @@ set(VCPKG_LIBRARY_LINKAGE static)
 
 # ...but with explicit platform toolset, so that future toolsets
 # aren't automatically picked up (it defaults to the latest one).
-set(VCPKG_PLATFORM_TOOLSET v143)
+set(VCPKG_PLATFORM_TOOLSET v145)
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply dep/vcpkg-overlay-triplets/x64-windows-static.cmake]: Apologies — that diff leaked from a local build-env workaround. Reverted to `v143` in 2e0375747. This PR does not touch the toolchain.

File: tools/wta/src/app.rs [hunk 8]
```diff
@@ -945,6 +951,11 @@ impl TabSession {
         self.selection_visible_pending = false;
         self.turn = TurnState::Idle;
         self.clear_recommendations();
+        // /clear, /new, /restart, tab reset and session-load all funnel
+        // through here. Queued prompts belonged to the conversation being
+        // wiped — survive across the reset would surprise the user by
+        // firing into the fresh session.
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Fixed in 2e0375747 — rephrased to `allowing them to survive across the reset would surprise...`.

File: tools/wta/src/ui/queued_hint.rs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    let mut out = String::new();
+    let mut used = 0usize;
+    for ch in text.chars() {
+        let w = UnicodeWidthChar::width(ch).unwrap_or(0);
+        if used + w > max_cells {
+            // Replace the last char with an ellipsis if there's room.
+            if !out.is_empty() {
+                out.pop();
+                out.push('…');
+            } else {
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Good catch — fixed in ee6e5b5c5. The truncation algorithm now reserves 1 cell upfront for the ellipsis (`content_budget = max_cells - 1`), so the result width is provably `<= max_cells` even with zero-width combining marks or wide CJK glyphs. Treats zero-width chars as 1 cell during budgeting (same convention as `ui/input.rs::char_display_width`). Added regression tests `truncate_respects_width_with_combining_marks` and `truncate_wide_char_with_two_cell_budget_emits_ellipsis_only`.

File: tools/wta/locales/en-US.yml [hunk 10]
```diff
@@ -62,6 +62,19 @@ input.placeholder.connecting: "connecting..."
...
+# waiting for the agent to finish the current turn. %{count} is the number
+# of queued prompts; %{preview} is a single-line preview of the most-recently
+# queued prompt (the one Esc would pop next).
+input.queue.indicator: "Queued (%{count}): %{preview}"
+# Transient hint shown after Esc removes the most-recently queued prompt.
+input.queue.removed: "Removed queued message: %{preview}"
+# Transient hint shown when the user tries to queue another prompt but the
+# queue is already at capacity (%{cap} entries). The user must wait for the
+# queue to drain or press Esc to remove queued items.
+input.queue.full: "Queue is full (%{cap}). Wait for the agent or press Esc to remove queued items."  # {Locked="Esc"}
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/locales/en-US.yml:76]: Translated to all 88 non-English locale files (including pseudo-locales) in 7d4c7d9e7. `%{count}`, `%{preview}`, `%{cap}`, and `Esc` are preserved verbatim per the `{Locked=Esc}` directive. Closes the localization gap.

File: tools/wta/src/ui/queued_hint.rs [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+use ratatui::prelude::*;
+use ratatui::widgets::Paragraph;
+
+use crate::app::App;
+use crate::theme;
+
+/// Height in rows the queued-hint occupies for `tab`. Zero when there's
+/// nothing to show — the layout collapses to the existing geometry.
+pub(crate) fn queue_hint_height(app: &App) -> u16 {
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Fixed in 8561e418b — reworded to 'for the current tab'.

File: tools/wta/src/ui/queued_hint.rs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    if max_cells == 0 {
+        return String::new();
+    }
+    // Fast path: fits as-is. Use the same width semantics for the check
+    // that we use during the truncation loop (zero-width → 1 cell), so a
+    // string with combining marks doesn't slip past this guard only to
+    // get clipped below.
+    if UnicodeWidthStr::width(text) <= max_cells {
+        return text.to_string();
+    }
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Fixed in 8561e418b. The fast-path comment was misleading — `UnicodeWidthStr::width` is actually the right check there because combining marks add 0 visible columns when overlaid on a base char, so they don't increase the rendered width. The 'zero-width → 1 cell' treatment inside the truncation loop is only there to budget room for per-char break decisions once we already know we must clip. Comment updated to explain that.

File: tools/wta/src/ui/queued_hint.rs [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        assert_eq!(truncate_to_width("hello", 10), "hello");
+    }
+
+    #[test]
+    fn truncate_over_width_inserts_ellipsis() {
+        let out = truncate_to_width("abcdefghij", 5);
+        // We push 5 chars then the next overflow triggers ellipsis swap.
+        assert!(out.ends_with('…'), "got: {out}");
+        assert!(out.chars().count() <= 5);
+    }
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs:136]: Fixed in 8561e418b. Test comment updated to match the reserve-1-for-ellipsis algorithm, and added a `UnicodeWidthStr::width(out) == 5` assertion to pin the contract going forward.

File: tools/wta/src/ui/queued_hint.rs [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+        preview = preview
+    )
+    .into_owned();
+    // Truncate again at the line level just in case the localized template
+    // expands beyond the available width (e.g. RTL or longer translations).
+    let truncated = truncate_to_width(&text, visible_width);
+    let line = Line::from(Span::styled(
+        format!("  {}", truncated),
+        theme::DIM,
+    ));
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Good catch — fixed in 5c073a2e1. The left padding is now clamped to `area.width - UnicodeWidthStr::width(truncated)` and capped at 2. In very narrow panes (`area.width < 2`) the padding shrinks to 0 so the truncation marker is always visible.

File: tools/wta/locales/en-US.yml [hunk 15]
```diff
@@ -62,6 +62,19 @@ input.placeholder.connecting: "connecting..."
 # Input box placeholder when agent is disconnected
 input.placeholder.disconnected: "disconnect"
 
+# ── Pending prompt queue (src/ui/queued_hint.rs, src/app.rs) ───────────────
+# Indicator shown above the input box while one or more prompts are queued
+# waiting for the agent to finish the current turn. %{count} is the number
+# of queued prompts; %{preview} is a single-line preview of the most-recently
+# queued prompt (the one Esc would pop next).
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/locales/en-US.yml:69]: Good catch — updated the PR description. The 'Out of scope' section now reflects that all 89 locales are translated; only chat-pane bidi reordering for RTL languages remains deferred.

**Comment by yeelam-gordon** [comment]
Round 6 follow-ups in 2f4ac63ef — addressed the three suppressed-low-confidence advisories from Copilot's last review:

1. **Single truncation pass**: `QueuedPrompt::preview` was clipping at char count + ellipsis; `queued_hint` was clipping again at cell width + ellipsis. Split: `collapsed_text()` for the indicator (whitespace-only collapse), `preview()` retained for transient hints. Avoids doubled '…' artefacts.
2. **Narrow-pane fix**: budget for `truncate_to_width` is now floored at 1 whenever `area.width >= 1`, so panes ≤2 cells still show the truncation marker instead of a blank row.
3. **Drain early-out**: `drain_pending_prompts` now does a cheap `any()` scan before allocating + sorting tab keys. The common 'no queues anywhere' tick is allocation-free.

284 tests pass.

File: tools/wta/src/ui/queued_hint.rs [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    //
+    // Use the raw collapsed text (no char-truncation) and let the single
+    // width-aware truncation pass below own all clipping. This avoids
+    // double-ellipsis artefacts when both `QueuedPrompt::preview` and
+    // `truncate_to_width` would clip the same string.
+    let preview = tab
+        .pending_prompts
+        .back()
+        .map(|p| p.collapsed_text())
+        .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs:48]: Fixed in 777a293c4. `QueuedPrompt` now caches a pre-collapsed string at enqueue time (`QueuedPrompt::new`). The helper `collapse_whitespace` streams chars into a single `String` with no intermediate `Vec`, so the per-frame UI render path just clones the cached field.

File: tools/wta/src/app.rs [hunk 17]
```diff
@@ -5005,6 +5098,106 @@ impl App {
...
+                continue;
+            }
+            if tab.pending_prompts.is_empty() {
+                continue;
+            }
+            let session_id = tab
+                .session_id
+                .clone()
+                .unwrap_or_else(|| DEFAULT_TAB_ID.to_string());
+            let queued = self
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Real bug — fixed in 777a293c4. The `DEFAULT_TAB_ID` fallback is now allowed only when the draining tab IS the focused one (where `tab_for_session(DEFAULT_TAB_ID)` resolves to `self.tab_id` correctly). Background-tab drains without a real ACP `session_id` are deferred — queue stays put until `SessionAttached` arrives, and the next event tick retries. Added `drain_holds_queue_when_tab_has_no_session_id` regression test.

File: tools/wta/src/ui/layout.rs [hunk 18]
```diff
@@ -259,9 +265,10 @@ pub fn input_cursor_position(app: &App, area: Rect) -> Option<Position> {
             Constraint::Min(0),
             hint_height,
             rec_hint_height,
+            queue_hint_height,
             Constraint::Length(input_height),
         ])
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/layout.rs:284]: Real desync — fixed in fd11aae04. `input_cursor_position` now replicates the same `reserved_below` computation and `chat_estimate.min(chat_max)` clamp `render` uses, so the chunk geometry matches on chat overflow. Also picked up the welcome-hint logic in the same pass (was a suppressed-low-confidence advisory in this same review) so the cursor doesn't drift by 1 row during first-run typing.

File: tools/wta/src/ui/queued_hint.rs [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+    };
+    let truncated = truncate_to_width(&text, budget);
+    // Clamp the left padding to whatever room is left after the truncated
+    // body — in very narrow terminals padding shrinks to 0 so the marker
+    // stays visible.
+    let prefix_width = (area.width as usize).saturating_sub(
+        unicode_width::UnicodeWidthStr::width(truncated.as_str()),
+    );
+    let prefix = " ".repeat(prefix_width.min(2));
+    let line = Line::from(Span::styled(
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs:75]: Fixed in fd11aae04 — `prefix_width.min(HORIZONTAL_PADDING as usize)` instead of the hard-coded 2.

File: tools/wta/src/app.rs [hunk 20]
```diff
@@ -1284,6 +1295,82 @@ pub struct App {
...
+        if self.collapsed.chars().count() <= max_chars {
+            self.collapsed.clone()
+        } else {
+            let taken: String = self
+                .collapsed
+                .chars()
+                .take(max_chars.saturating_sub(1))
+                .collect();
+            format!("{}…", taken)
+        }
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs:1452]: Fixed in 689ab47c7. `preview(0)` now short-circuits to `String::new()`. Added `queued_prompt_preview_zero_budget_is_empty` regression test.

File: tools/wta/src/ui/queued_hint.rs [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    if app.current_tab().pending_prompts.is_empty() {
+        0
+    } else {
+        1
+    }
+}
+
+/// Width budget for the preview text inside the hint row. Mirrors the
+/// layout's left/right horizontal padding (1 cell each).
+const HORIZONTAL_PADDING: u16 = 2;
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Fixed in 689ab47c7 — reworded to `Two-cell left indent that the hint row prepends, matching the visual alignment of other transient-hint rows…`.

File: tools/wta/src/app.rs [hunk 22]
```diff
@@ -4243,6 +4328,42 @@ impl App {
...
             }
+            // Pop the most-recently queued prompt (LIFO undo) — runs BEFORE
+            // the in-flight cancel below so users can rescind a prompt they
+            // queued by mistake while the agent (or autofix) is still
+            // working on the head. Only when the input box is empty;
+            // otherwise the user is most likely clearing a draft.
+            KeyCode::Esc
+                if self.current_tab().input.is_empty()
+                    && !self.current_tab().pending_prompts.is_empty() =>
+            {
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs:4358]: Fixed in ba07958a4 — PR description rewritten to match the actual behavior (queue-pop first, then cancel-head). Added a 'Why queue-pop beats cancel-head' rationale explaining the autofix-with-queue scenario that drove the priority choice.

File: tools/wta/src/app.rs [hunk 23]
```diff
@@ -5758,6 +6013,46 @@ impl App {
...
     /// dropped by the stale-check in `turn_observe_chunk`.
+    /// Shared in-flight-turn cancel used by Esc and Ctrl+C. Mirrors `/stop`:
+    /// sends `session/cancel` to the ACP client, transitions the state
+    /// machine back to Idle, and pushes a "Cancelled." system bubble.
+    /// Returns `true` iff a turn was actually in flight and was cancelled.
+    ///
+    /// Note: when the queue is non-empty, the per-tick `drain_pending_prompts`
+    /// hook will promote the next queued prompt on the following tick. That's
+    /// the intended "Esc peels prompts off the head one by one" UX — each
+    /// Esc cancels the current head; the queue keeps rotating.
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Fixed in ba07958a4 — `cancel_in_flight_turn` doc comment rewritten to reflect that queue-pop runs BEFORE this helper, so Esc only cancels the head when `pending_prompts` is empty (or the call comes from Ctrl+C, which has no queue-pop short-circuit).

File: tools/wta/src/app.rs [hunk 24]
```diff
@@ -7961,4 +8264,395 @@
...
+        type_text(&mut app, "queued");
+        press_enter(&mut app);
+        assert_eq!(app.current_tab().pending_prompts.len(), 1);
+        app.current_tab_mut().clear_chat_history();
+        assert!(app.current_tab().pending_prompts.is_empty(),
+            "reset paths must not let stale prompts fire into a fresh session");
+    }
+
+    #[test]
+    fn drain_noops_when_disconnected() {
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Fixed in ba07958a4 — renamed to `drain_no_ops_when_disconnected`.

File: tools/wta/src/app.rs [hunk 25]
```diff
@@ -7961,4 +8264,395 @@
...
+        // Simulate transport drop.
+        app.state = ConnectionState::Disconnected;
+        app.drain_pending_prompts();
+        assert_eq!(app.current_tab().pending_prompts.len(), 1,
+            "drain must not fire while disconnected — the prompt would never reach the agent");
+    }
+
+    #[test]
+    fn queued_prompt_preview_truncates_and_collapses_whitespace() {
+        let q = QueuedPrompt::new("  hello\n\nworld   ".into());
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Fixed in ba07958a4 — adjusted the test fixture to `\\\\  hello  \n\n  world   \\\\\` so the spell checker no longer reads the `\n` escape + `world` as a single token `nworld`.

File: tools/wta/src/app.rs [hunk 26]
```diff
@@ -4203,12 +4284,16 @@ impl App {
...
             KeyCode::Esc
-                if self.current_tab().turn.recommendations().is_some()
-                    || (self.current_tab().autofix.pane_id.is_some()
-                        && !self.current_tab().turn.is_idle()) =>
+                if self.current_tab().turn.recommendations().is_some() =>
             {
-                // Dismiss armed fix card or cancel in-flight autofix request.
                 // `turn_cancel` bumps generation, emits autofix_state_cleared,
                 // and resets the state machine to Idle.
                 let session_id = self.current_tab().session_id.clone();
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Fixed in 1ccd674d1. The no-session-id branch now falls back to `DEFAULT_TAB_ID` (same convention `cancel_in_flight_turn` uses), so `turn_cancel` runs and properly resets the turn/dismisses the card in both branches. The previous manual-cleanup path was dead code for the card case since it didn't touch `tab.turn`.

File: tools/wta/src/ui/queued_hint.rs [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+        .pending_prompts
+        .back()
+        .map(|p| p.collapsed_text())
+        .unwrap_or_default();
+    let text = t!(
+        "input.queue.indicator",
+        count = count,
+        preview = preview
+    )
+    .into_owned();
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs:54]: Fixed in e2917ce19. `QueuedPrompt::collapsed` is now capped at `COLLAPSED_PREVIEW_CAP = 256` chars (well over the ~60-cell visible indicator). The new `collapse_whitespace_capped` helper breaks out of the scan as soon as the cap is hit, so a paste-large prompt costs O(cap) per UI frame instead of O(n). Full text remains intact on `QueuedPrompt::text` for ACP dispatch — only the rendering cache is capped. Regression test: `queued_prompt_caps_collapsed_storage`.

File: tools/wta/src/ui/queued_hint.rs [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+        assert_eq!(out, "…", "got: {out:?}");
+    }
+
+    #[test]
+    fn truncate_respects_width_with_combining_marks() {
+        // Zero-width combining marks can't push the visible width past
+        // `max_cells`. Regression for the Copilot round-3 review case:
+        // `a\u{0301}b` with `max_cells=1` previously emitted `a…` which
+        // displays as 2 cells.
+        let out = truncate_to_width("a\u{0301}bcdef", 1);
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/ui/queued_hint.rs]: Fixed in 97c9e597a — swapped `\\\cdef\\\\` filler for `\\\1234\\\\`. Digits don't trip `check-spelling`, and the test contract is unchanged: one grapheme `á` followed by *something* past `max_cells=1` to verify combining-mark width handling. Added a comment explaining the choice so future maintainers don't wonder either.

File: tools/wta/src/app.rs [hunk 29]
```diff
@@ -1379,6 +1390,107 @@ pub struct App {
...
+    /// `ui/queued_hint::render`. Avoids both the per-frame
+    /// `split_whitespace().collect::<Vec<_>>().join(" ")` allocation and
+    /// unbounded scanning/copying when the user pastes a huge prompt —
+    /// the visible indicator can only show ~60 cells anyway.
+    collapsed: String,
+}
+
+/// Upper bound on `QueuedPrompt::collapsed`. Generous over the ~60-cell
+/// visible indicator (`PREVIEW_MAX_CHARS` in `ui/queued_hint`), but caps
+/// per-frame `t!()` interpolation + width scanning to a constant when the
```
[the remark this answers is not in the document]
Comment (yeelam-gordon) [reply tools/wta/src/app.rs]: Good catch — fixed in d94f754fe. Removed the stale reference to `PREVIEW_MAX_CHARS` (which I deleted in an earlier review round when switching to width-based truncation). Comment now clarifies the cap is an O(n) defense against unbounded per-frame work, not a display cap — the actual visible width is enforced by `truncate_to_width` against `area.width`.

**Comment by yeelam-gordon** [comment]
The agent pane need more experience changed so that this "queue" can be working good.
e.g. how can I focus back to previous suggestion.
